subprojects/core/src/main/java/org/gradle/configuration/ConfigurationTargetIdentifier.java

```diff
@@ -0,0 +1,127 @@
+/*
+ * Copyright 2017 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.configuration;
+
+import org.gradle.api.Nullable;
+import org.gradle.api.internal.GradleInternal;
+import org.gradle.api.internal.SettingsInternal;
+import org.gradle.api.internal.plugins.PluginAwareInternal;
+import org.gradle.api.internal.project.ProjectInternal;
+
+/**
+ * Uniquely identifies the target of some configuration.
+ *
+ * This is primarily used to support
+ * {@code ApplyScriptPluginBuildOperationType.Details} and {@code ApplyPluginBuildOperationType.Details}.
+ */
+public abstract class ConfigurationTargetIdentifier {
+
+    private ConfigurationTargetIdentifier() {
+    }
+
+    public enum Type {
+        GRADLE,
+        SETTINGS,
+        PROJECT;
+
+        public final String label = name().toLowerCase();
+    }
+
+    // One of "gradle", "settings, project
+    abstract public Type getTargetType();
+
+    @Nullable
+    // for Gradle or Settings, and unknown
```

Comment:
This reads weirdly, probably should be a proper javadoc comment like on the interface.